src/prop/zero_level_learner.h

```diff
@@ -0,0 +1,86 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Learner for literals asserted at level zero.
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__PROP__ZERO_LEVEL_LEARNER_H
+#define CVC5__PROP__ZERO_LEVEL_LEARNER_H
+
+#include <unordered_set>
+
+#include "context/cdhashset.h"
+#include "context/cdo.h"
+#include "expr/node.h"
+#include "smt/env_obj.h"
+
+namespace cvc5 {
+namespace prop {
+
+class PropEngine;
+
+/**
+ * The proxy class that allows the SatSolver to communicate with the theories
+ */
+class ZeroLevelLearner : protected EnvObj
+{
+  using NodeSet = context::CDHashSet<Node>;
+
+ public:
+  ZeroLevelLearner(Env& env, PropEngine* propEngine);
+
+  ~ZeroLevelLearner();
+
+  void notifyInputFormulas(const std::vector<Node>& assertions,
+                           std::unordered_map<size_t, Node>& skolemMap,
+                           const std::vector<Node>& ppl);
+  /**
+   * Notify the given literal was asserted
+   */
+  void notifyAsserted(TNode assertion);
+
+  /** Get the zero-level assertions */
+  const std::unordered_set<Node>& getLearnedZeroLevelLiterals() const;
+
+ private:
+  static void getAtoms(TNode a,
+                       std::unordered_set<TNode>& visited,
+                       std::unordered_set<TNode>& ppLits);
+
+  /** The prop engine we are using. */
+  PropEngine* d_propEngine;
+
+  /** Set of literals that hold at level 0 */
+  std::unordered_set<TNode> d_levelZeroAsserts;
+
+  /** Set of learnable literals that hold at level 0 */
+  std::unordered_set<Node> d_levelZeroAssertsLearned;
+
+  /** Whether we have seen an assertion level > 0 */
+  context::CDO<bool> d_nonZeroAssert;
+
+  /** Preprocessed literals that are not learned */
+  std::unordered_set<TNode> d_ppnAtoms;
+
+  /** Already learned TEMPORARY */
+  std::unordered_set<TNode> d_pplAtoms;
+
+  /** Current counter of assertions */
+  size_t d_assertNoLearnCount;
```

Comment:
Is this counter used anywhere?

Reply:
It is used on a dev branch that uses this utility.